chain/ethhashlookup/eth_transaction_hash_lookup.go

```diff
@@ -78,31 +129,43 @@ func (ei *EthTxHashLookup) GetHashFromCid(c cid.Cid) (ethtypes.EthHash, error) {
 }
 
 func (ei *EthTxHashLookup) DeleteEntriesOlderThan(days int) (int64, error) {
-	res, err := ei.db.Exec("DELETE FROM eth_tx_hashes WHERE insertion_time < datetime('now', ?);", "-"+strconv.Itoa(days)+" day")
-	if err != nil {
-		return 0, err
+	if ei.db == nil {
+		return 0, xerrors.New("db closed")
 	}
 
-	return res.RowsAffected()
-}
-
-func NewTransactionHashLookup(ctx context.Context, path string) (*EthTxHashLookup, error) {
-	db, _, err := sqlite.Open(path)
+	res, err := ei.stmtDeleteOlderThan.Exec("-" + strconv.Itoa(days) + " day")
 	if err != nil {
-		return nil, xerrors.Errorf("failed to setup eth transaction hash lookup db: %w", err)
-	}
-
-	if err := sqlite.InitDb(ctx, "eth transaction hash lookup", db, ddls, []sqlite.MigrationFunc{}); err != nil {
-		_ = db.Close()
-		return nil, xerrors.Errorf("failed to init eth transaction hash lookup db: %w", err)
+		return 0, err
 	}
-
-	return &EthTxHashLookup{db: db}, nil
+	return res.RowsAffected()
 }
 
-func (ei *EthTxHashLookup) Close() error {
+func (ei *EthTxHashLookup) Close() (err error) {
 	if ei.db == nil {
 		return nil
 	}
-	return ei.db.Close()
+	// set the db to nil first to ensure further queries or inserts fail early,
+	// then close preparedstatements, then close the db
+	db := ei.db
+	defer func() {
+		err = db.Close()
+	}()
+	ei.db = nil
+	if ei.stmtInsertTxHash != nil {
+		_ = ei.stmtInsertTxHash.Close()
+		ei.stmtInsertTxHash = nil
```

Comment:
Like if there really is concurrent access to both  this `Close` function and DB write operations, this can cause a nil panic if someone tries to use this after it's set to nil as we have no synchronization here.